include/tvm/expr.h

```diff
@@ -148,34 +219,52 @@ class Integer : public Expr {
   using ContainerType = IntImm;
 };
 
+/*! \brief range over one dimension */
+class RangeNode : public Node {
+ public:
+  /*! \brief beginning of the node */
+  Expr min;
+  /*! \brief the extend of range */
+  Expr extent;
+  /*! \brief constructor */
+  RangeNode() {}
+  RangeNode(Expr min, Expr extent) : min(min), extent(extent) {}
 
-/*! \brief container class of iteration variable. */
-class IterVarNode;
+  void VisitAttrs(AttrVisitor* v) final {
+    v->Visit("min", &min);
+    v->Visit("extent", &extent);
+  }
 
-/*!
- * \brief same as HalideIR::IR::Range
- *  except it provide an constructor with (begin, end)
- *
- *  \note Traditional Halide's Range have a constructor with
- *   (begin, extent), which does not match the convention in e.g. python.
- *   We decided to correct it by removing the constructor in HalideIR,
- *   and add it back in TVM's range.
- */
-class Range : public HalideIR::IR::Range {
+  static constexpr const char* _type_key = "Range";
+  TVM_DECLARE_NODE_TYPE_INFO(RangeNode, Node);
+};
+
+/*! \brief Range constainer  */
+class Range : public NodeRef {
  public:
-  /*! \brief constructor */
-  Range() {}
-  explicit Range(NodePtr<Node> n) : HalideIR::IR::Range(n) {}
   /*!
    * \brief constructor by begin and end
    * \param begin The begin of the range.
    * \param end The end of the range.
    */
   TVM_DLL Range(Expr begin, Expr end);
-
-  TVM_DLL static Range make_by_min_extent(Expr min, Expr extent);
+  /*!
+   * \brief construct a new range with min and extent
+   *  The corresponding constructor is removed,
+   *  because that is counter convention of tradition meaning
+   *  of range(begin, end)
+   *
+   * \param min The minimum range.
+   * \param extent The extent of the range.
+   */
+  static Range make_by_min_extent(Expr min, Expr extent);
+  // declare range.
+  TVM_DEFINE_NODE_REF_METHODS(Range, NodeRef, RangeNode);
```

Comment:
Can we unify the RELAY and TVM macros into one?

Reply:
I agree, likely we will need to use these macros instead in relay and remove relay specific macros.